PrimeString Solution

arjun9218 opened the pull request:
Prime String solution also includes JUnit test

arjun9218 committed:
PrimeString Solution

Prime String solution also includes JUnit test

bsmaheshkumar commented on Solutions-1/PrimeString.java:
```diff
@@ -0,0 +1,118 @@
+		logger = Logger.getLogger(PrimeString.class.getSimpleName());
+	}
+	
+	/**
```
which code formate is applied?

arjun9218 replied:
I had manually entered the comments, I had not applied code format.

bsmaheshkumar reviewed (commented):
Thanks @arjun9218  for solving this problem.
As part of Review request, could you please include the screen shots of pass test cases and test code coverage?

bsmaheshkumar commented on Solutions-1/PrimeString.java:
```diff
@@ -0,0 +1,118 @@
+	 * @return true, if is prime
+	 */
+	
+	protected boolean isPrime(Integer number) {
```
which this method is protected, why not public of private?

arjun9218 replied:
Yeah sure I'll convert this to maven project and solve two more problems.

bsmaheshkumar commented on Solutions-1/PrimeString.java:
```diff
@@ -0,0 +1,118 @@
+	 * @return true, if is Prime String
+	 */
+	
+	protected String isPrimeString(String input) {
```
Only this method can be public rest can be private.

arjun9218 replied:
I had made that method as protected because I wanted to test all the methods, I read your review comments and I'll make all the methods private and the isPrimeString method as public.

bsmaheshkumar replied:
Ok.
You can also unit test private method's view public method.
No need to change the access of the method only for unit testing.

arjun9218 replied:
Ok got it.

bsmaheshkumar commented on Solutions-1/PrimeStringTest.java:
```diff
@@ -0,0 +1,55 @@
+	 */
+	
+	@Before
+	public void preRequisites() {
```
Why do we need this test cases?

arjun9218 replied:
I thought of testing all the methods so I had put this. I'll make changes as per your comments.

bsmaheshkumar commented on Solutions-1/PrimeStringTest.java:
```diff
@@ -0,0 +1,55 @@
+	 */
+	
+    @Test
+    public void testing() { 
```
what about Negative test case?
What is the unit test coverage?

bsmaheshkumar reviewed (changes requested):
Need to add more unit test cases, both positive and negative.
Please address the review comments.
Can you convert the same to maven project? And solve two more problem? Before we move on to Angular?

arjun9218 committed:
Four Solutions for set 1 problems

Solutions are backed up by unit tests. Most of the tests cover positive and negative test cases.

arjun9218 committed:
Snapshot containing test runs

All tests are successful

arjun9218 committed:
Added screenshot of Code coverage

bsmaheshkumar commented:
@neerajkujha @coderhindu 
Adding Neeraj and Narasimhan to review the code.

bsmaheshkumar commented on solutuins-1/pom.xml:
```diff
@@ -0,0 +1,27 @@
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>solutions-1</groupId>
+  <artifactId>solutuins-1</artifactId>
```
typo solutions

bsmaheshkumar commented on solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:
```diff
@@ -0,0 +1,52 @@
+		try {
+			int evenWeaponsCount = 0;
+			int oddWeaponsCount = 0;
+			for(int i=0;i<size;i++) {
```
Can you apply code format.

arjun9218 replied:
Sure I'll apply code format and commit the latest code

bsmaheshkumar commented on solutuins-1/src/main/java/solutions_1/solutuins_1/PrimeString.java:
```diff
@@ -0,0 +1,118 @@
+	 * @return true, if is Prime String
+	 */
+	
+	private boolean isPrimeStringLogic(Map<Character, Integer> map) {
```
do we need logic in method name, isPrimeString should be enough?

arjun9218 replied:
There is one more method isPrimeString, isPrimeStringLogic is just a supporting method.

bsmaheshkumar commented on solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:
```diff
@@ -0,0 +1,52 @@
+	 * @return String ready or not ready for battle based on Kattapa's luck logic
+	 */
+
+	public String isReadyOrNot(int size, Integer[] soldierWeapons) {
```
Can this method be renamed to isReadyForBattle
On need to Use is and or together in function name.

arjun9218 replied:
Yeah. I'll rename it

bsmaheshkumar commented on solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:
```diff
@@ -0,0 +1,61 @@
+	 * Instantiating KattappaTest object
+	 */
+
+	KattapaReadyForBattle instance = new KattapaReadyForBattle();
```
can the instance be rename to kattapaBattleTest?

bsmaheshkumar commented on solutuins-1/src/test/java/solutions_1/solutuins_1/KattappaTest.java:
```diff
@@ -0,0 +1,61 @@
+	 */
+
+	@Test
+	public void testing() { 
```
The method name code be like kattapaReadyForBattle_Positive_Test

The verification can be simple like below
```
@Test
public void kattapaReadyForBattle_Positive_Test  () {
   assertEquals("NOT READY", kattapaBattleTest.isReadyForBattle(4, new Integer[]{11, 12, 13, 14}));
   assertEquals("READY FOR BATTLE", kattapaBattleTest.isReadyForBattle(4, new Integer[]{2, 3, 4));
}
```

bsmaheshkumar replied:
Similar to Positive Add Negative test cases

```
@Test
public void kattapaReadyForBattle_Negative_Test  () {
   assertEquals("Number of Soliders and weapon not matching ", kattapaBattleTest.isReadyForBattle(2, new Integer[]{11, 12, 13, 14}));
   assertEquals("Invalid input", kattapaBattleTest.isReadyForBattle(4, null);
}
```

Please include the junit result image.
And also include test code coverage image.

arjun9218 replied:
I had included a screenshot. It's a .bmp file.

bsmaheshkumar commented on solutuins-1/src/test/java/solutions_1/solutuins_1/NonReducibleTest.java:
```diff
@@ -0,0 +1,59 @@
+
+
+@RunWith(Parameterized.class)
+public class NonReducibleTest {
```
Similar to KattapaReadyForBattle test class, please update the tests other classes.

arjun9218 replied:
Sure, will do and commit the changes

bsmaheshkumar commented on solutuins-1/src/test/java/solutions_1/solutuins_1/ShoppingTest.java:
```diff
@@ -0,0 +1,59 @@
+	public void testing() { 
+		int result = instance.getRoyPayment( shoppingCart );
+		//String expected = "Yes";
+		System.out.println("Expected : " + expected + " | Actual : " + result);
```
do we need sys out?

arjun9218 replied:
I was trying to make the Test runs to look little detailed. I had attached snapshots to base directory, I've again added screenshots in a separate folder.

bsmaheshkumar reviewed (changes requested):
Few changes in unit test cases required.

arjun9218 committed:
Screenshots of test runs and test coverage

arjun9218 committed:
Latest commit addressing review comments

Added few more test cases. Changed the naming conventions and applied code format. Please find the code coverage and test runs screenshots in "Screen snaps" folder

arjun9218 commented:
@simha111 adding Narasimha for code review

neerajkujha commented on Solutions-1/src/main/java/solutions_1/solutions_1/PrimeString.java:
```diff
@@ -0,0 +1,121 @@
+			} else {
+				return "No";
+			}
+		} catch (Exception e) {
```
what kind of exception do you expect to be thrown?

arjun9218 replied:
I was thinking if the input was null, then NullPointerException may get thrown so I had surrounded the code with try catch.

neerajkujha replied:
It is always better to handle null or empty cases in code than throwing exceptions. As name suggests exceptions should be used for exceptional situations.  Handling and validating input should be developer responsibility and handled more elegantly and efficiently

arjun9218 replied:
Hmmm yeah I agree with that. I'll keep that in mind.

neerajkujha commented on solutuins-1/src/main/java/solutions_1/solutuins_1/KattapaReadyForBattle.java:
```diff
@@ -0,0 +1,52 @@
+	 * @return String ready or not ready for battle based on Kattapa's luck logic
+	 */
+
+	public String isReadyOrNot(int size, Integer[] soldierWeapons) {
```
why an array of Integer not int?

arjun9218 replied:
No particular reason here, can use int as well. In PrimeString problem I had used Maps which didn't accept the primitive datatype int so I was using Integer there. In that flow I had used Integer here.

neerajkujha replied:
Can you please explain difference between both with advantages and disadvantages. Cant we use int in map directly instead of Integer?

arjun9218 replied:
Integer is a wrapper class right, so it has many additional methods. Primitive datatypes are just basic. Advantage of a primitive datatype is that performance wise it is faster compared to Wrapper classes, and wrapper classes takes up more space compared to primitive datatypes.
We can't use int directly in the map because it accepts only classes derived from Object, int, float etc., primitives are not derived from Object.

neerajkujha commented on solutuins-1/src/main/java/solutions_1/solutuins_1/NonReducible.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * The Class ChopCount.
+ */
+class ChopCount {
```
why we need a class here? Can we achieve this without this class?

arjun9218 replied:
In the problem statement, the expected output was an object with reduced string and number of operation so I had created such class. We can just return an anonymous class Object like /* Object obj = new Object(){ //reduced string and number of operations variables}; */

neerajkujha replied:
Can you please explain this anonymous class object.
Not able to comprehend your explanation.

arjun9218 replied:
![anonymous_obj](https://user-images.githubusercontent.com/3712951/41608370-02a7466a-7406-11e8-80b4-f191cfd76c81.png)
Hi Neeraj,

Please find the attached image for the object that we can return instead of creating a ChopCount class.

Regards,
Arjun

neerajkujha replied:
Thanks

arjun9218 replied:
One more way is to just have those variables in NonReducible class itself and return the same instead of having a separate ChopCount class.
